### PR TITLE
Block service

### LINKED ISSUE #449: BlockService for merkle auditing
Adding a new `BlockService` to validate a given merkle path item against the root hash at a specific block height. 

e.g. `validateInBlock(...)`

### PR DESCRIPTION
**Pull request by rg911**
Added `BlockService` to validate transaction and receipt hashes against in specific block

Fixed #449

### CODE REVIEW

File: src/service/BlockService.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        }
+        const rootToCompare = merklePathItem.reduce((proofHash, pathItem) => {
+                const hasher = sha3_256.create();
+                // Left
+                if (pathItem.position === 1) {
+                    return hasher.update(Buffer.from(pathItem.hash + proofHash, 'hex')).hex();
+                } else {
+                    // Right
+                    return hasher.update(Buffer.from(proofHash + pathItem.hash, 'hex')).hex();
+                }
```
Comment (fboucquez) [inline src/service/BlockService.ts]: The position is being changed from a number to a (left/right) string. Rest has done the change in master, we are still waiting for the open api PR https://github.com/nemtech/nem2-openapi/pull/77

@dgarcia360 can we move 77 PR forward? once released we can then generate the clients and update the ts and java PR related to merkle tests

File: src/service/BlockService.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+     * @param leaf transaction hash
+     * @param height block height
+     */
+    public validateTransactionInBlock(leaf: string, height: UInt64): Observable<boolean> {
+        const rootHashObservable = this.blockRepository.getBlockByHeight(height);
+        const merklePathItemObservable = this.blockRepository.getMerkleTransaction(height, leaf);
+        return combineLatest(rootHashObservable, merklePathItemObservable).pipe(
+            map((combined) => this.validateInBlock(leaf, combined[1].merklePath, combined[0].blockTransactionsHash)),
+        );
+    }
```
Comment (fboucquez) [inline src/service/BlockService.ts]: When rest raises an error, like block, transaction or recipients not found, should this method catch the exception and return false or just raise the exception?

I think it should raise the error but no 100% sure. @rg911 @dgarcia360

**Comment by fboucquez** [comment]
@rg911 can we add unit testing to the block service. E2E tests are great but they don't run every build and they don't increase coverage.

File: src/service/BlockService.ts (hunk 2 above)
Comment (rg911) [reply src/service/BlockService.ts]: In fact if the transaction / receipt doesn't exist. The exception should be captured before the validation within other http repos

Comment (fboucquez) [reply src/service/BlockService.ts]: I think adding` .pipe(catchError(error => of(false)))` would do the trick after combineLatest

File: src/service/BlockService.ts (hunk 1 above)
Comment (dgarcia360) [reply src/service/BlockService.ts]: https://github.com/nemtech/nem2-openapi/releases/tag/v0.8.4

File: e2e/service/BlockService.spec.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                expect(validationResult).to.be.true;
+            } else {
+                assert(false, `Transaction (hash: ${transactionHash}) not found`);
+            }
+        });
+    });
+
+    describe('Validate receipt', () => {
+        it('call block service', async () => {
+            const statement = await receiptRepository.getBlockReceipts(UInt64.fromUint(1)).toPromise();
```
Comment (dgarcia360) [inline e2e/service/BlockService.spec.ts]: const statement**s**

File: e2e/service/BlockService.spec.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            const receipt = statement.transactionStatements[0];
```
Comment (dgarcia360) [inline e2e/service/BlockService.spec.ts]: const ~receipt~ statement

File: src/service/BlockService.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * Validate receipt hash in block
+     * @param leaf receipt hash
+     * @param height block height
+     */
+    public validateReceiptInBlock(leaf: string, height: UInt64): Observable<boolean> {
```
Comment (dgarcia360) [inline src/service/BlockService.ts]: Should the function be named ``validateStatementInBlock``? The hash passed as leaf belongs to a transaction/resolution statement, not to an individual receipt.

**Comment by fboucquez** [comment]
@rg911 0.8.4-SNAPSHOT.202002181344 client has been pushed

**Comment by rg911** [comment]
Rebased to #453

**Pull request closed without merging**